A config option for using enterprise record format in exceptional cases.

maxsumrall opened the pull request:
We had managed to do this without using a config option, but when trying to do recovery-robustness on both record formats this new config seems to be the path of least resistance. This PR only introduces this config option, copying how we do it with lock managers.

Commit:
A config option for using enterprise record format in exceptional cases.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/InternalRecordFormatSelector.java:
```diff
@@ -28,13 +34,34 @@
+            else if ( key.equals( "" ) )
+            {
+                logging.getInternalLog( CommunityFacadeFactory.class )
+                        .info( "No locking implementation specified, defaulting to '" + candidateId + "'" );
```
Should be `record format` instead of `locking implementation`.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/InternalRecordFormatSelector.java:
```diff
@@ -28,13 +34,34 @@
+        {
+            return LowLimit.RECORD_FORMATS;
+        }
+        else if ( key.equals( "" ) )
```
mmh, same here about the empty string

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -90,6 +90,7 @@
 
         // Kept here to have it not be publicly documented.
         public static final Setting<String> lock_manager = setting( "lock_manager", Settings.STRING, "" );
+        public static final Setting<String> record_format = setting( "record_format", Settings.STRING, "" );
```
Doesn't this default mean that you have to explicitly specify "highlimit" to get that format? Or that's just temporary?

maxsumrall replied:
If the high limit format is available it will choose it unless you specify "community". The empty string as well as "highlimit" will result in using the high limits if it is available.

spacecowboy replied:
Yes, that is caught by added tests below

tinwelint replied:
OK great

spacecowboy commented:
Some more tests for `HighLimitTest.java`

``` java
    @Test
    public void shouldResolveCommunityRecordFormat() throws Exception
    {
        Config config = new Config( MapUtil.stringMap( "record_format", "community" ) );
        RecordFormats formatSelector = InternalRecordFormatSelector.select( config, NullLogService.getInstance() );
        assertEquals( LowLimit.RECORD_FORMATS.storeVersion(), formatSelector.storeVersion() );
    }

    @Test
    public void shouldResolveNoRecordFormatToHighLimitDefault() throws Exception
    {
        Config config = new Config();
        RecordFormats formatSelector = InternalRecordFormatSelector.select( config, NullLogService.getInstance() );
        assertEquals( HighLimit.RECORD_FORMATS.storeVersion(), formatSelector.storeVersion() );
    }

    @Test
    public void shouldNotResolveNoneExistingRecordFormat() throws Exception
    {
        Config config = new Config( MapUtil.stringMap( "record_format", "superawesomeimaginaryformat" ) );
        try
        {
            RecordFormats formatSelector = InternalRecordFormatSelector.select( config, NullLogService.getInstance() );
            fail( "Should not be possible to specify non-existing format" );
        }
        catch ( IllegalArgumentException ignored )
        {
            // Success
        }
    }
```

spacecowboy commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/PropertyStore.java:
```diff
@@ -72,11 +72,10 @@ public PropertyStore(
             DynamicArrayStore arrayPropertyStore,
-            RecordFormat<PropertyRecord> recordFormat,
-            String storeVersion )
+           RecordFormats recordFormats)
```
Inconsistent indentation

maxsumrall replied:
✔️

Commit:
Fixes and more tests for config option.

Had missed a case to return high limits format when it should have been there,
and added tests from Jonas which test for that.

Reverted a change in TransactionRecordStateTest which caused config
options to not be applied correctly.

tinwelint commented:
@maxsumrall failing on `HighLimitTest` which I assume is now because the actual selection is disabled (until migration is in). Please ignore those tests and make a note on store migration work to enabled again.

maxsumrall commented:
Green and there is a note on store migration card to take care of the `HighLimitTest` when that work is done.
